File: internal/domain/auth/model/normalize.go
package model

import (
	"fmt"
	"strings"

	"github.com/samber/lo"

	"github.com/rendau/ruto/internal/constant"
)

func (m *Auth) Normalize() error {
	m.Mode = strings.ToLower(strings.TrimSpace(m.Mode))
	if m.Mode == "" {
		m.Mode = constant.AuthModeExtend
	}
	if !constant.AuthModeIsValid(m.Mode) {
		return fmt.Errorf("mode: is invalid")
	}

	if m.Methods == nil {
		m.Methods = []*AuthMethod{}
	}
	for i := range m.Methods {
		if err := m.Methods[i].Normalize(); err != nil {
			return fmt.Errorf("methods[%d]: %w", i, err)
		}
	}

	return nil
}

func (m *AuthMethod) Normalize() error {
	var methodsCount int

	if m.Basic != nil {
		methodsCount++
		if err := m.Basic.Normalize(); err != nil {
			return fmt.Errorf("basic: %w", err)
		}
	}
	if m.APIKey != nil {
		methodsCount++
		if err := m.APIKey.Normalize(); err != nil {
			return fmt.Errorf("api_key: %w", err)
		}
	}
	if m.JWT != nil {
		methodsCount++
		if err := m.JWT.Normalize(); err != nil {
			return fmt.Errorf("jwt: %w", err)
		}
	}
	if m.IPValidation != nil {
		methodsCount++
		if err := m.IPValidation.Normalize(); err != nil {
			return fmt.Errorf("ip_validation: %w", err)
		}
	}

	if methodsCount == 0 {
		return fmt.Errorf("empty")
	}

	return nil
}

func (m *AuthMethodBasic) Normalize() error {
	for i := range m.Users {
		if err := m.Users[i].Normalize(); err != nil {
			return fmt.Errorf("users[%d]: %w", i, err)
		}
	}
	return nil
}

func (m *AuthMethodBasicUser) Normalize() error {
	m.Username = strings.TrimSpace(m.Username)
	if m.Username == "" {
		return fmt.Errorf("username: empty")
	}

	m.Password = strings.TrimSpace(m.Password)

	return nil
}

func (m *AuthMethodAPIKey) Normalize() error {
	m.Header = strings.TrimSpace(m.Header)
	if m.Header == "" {
		m.Header = "X-Api-Key"
	}
	m.Keys = lo.FilterMap(m.Keys, func(v AuthMethodAPIKeyItem, _ int) (AuthMethodAPIKeyItem, bool) {
		v.Name = strings.TrimSpace(v.Name)
		v.Key = strings.TrimSpace(v.Key)
		return v, v.Key != ""
	})
	if len(m.Keys) == 0 {
		return fmt.Errorf("keys: empty")
	}
	return nil
}

func (m *AuthMethodJWT) Normalize() error {
	m.Kid = strings.TrimSpace(m.Kid)
	if m.Kid == "" {
		return fmt.Errorf("kid: empty")
	}
	m.Roles = lo.FilterMap(m.Roles, func(v string, _ int) (string, bool) {
		v = strings.TrimSpace(v)
		return v, v != ""
	})
	m.Roles = lo.Uniq(m.Roles)
	return nil
}

func (m *AuthMethodIPValidation) Normalize() error {
	m.AllowedIps = lo.FilterMap(m.AllowedIps, func(v AuthMethodIPValidationItem, _ int) (AuthMethodIPValidationItem, bool) {
		v.Name = strings.TrimSpace(v.Name)
		v.Ip = strings.TrimSpace(v.Ip)
		return v, v.Ip != ""
	})
	if len(m.AllowedIps) == 0 {
		return fmt.Errorf("allowed_ips: empty")
	}
	return nil
}

File: apps/admin/src/api/normalize.ts
import type {
  AppMain,
  Auth,
  AuthMethod,
  AuthMethodApiKey,
  AuthMethodApiKeyItem,
  AuthMethodBasic,
  AuthMethodBasicUser,
  AuthMethodIpValidation,
  AuthMethodIpValidationItem,
  AuthMethodJwt,
  EndpointMain,
  Logging,
  RootMain,
  Variable
} from "./types";

export function emptyAuth(): Auth {
  return { enabled: false, mode: "extend", methods: [] };
}

export function emptyLogging(): Logging {
  return {
    mode: "extend",
    level: "",
    headers: false,
    query_params: false,
    req_body: false,
    resp_body: false,
    req_body_limit: 0,
    resp_body_limit: 0
  };
}

export function normalizeLogging(value?: Logging | null): Logging {
  if (!value) {
    return emptyLogging();
  }
  const level =
    value.level === "all" || value.level === "error" || value.level === "none" ? value.level : "";
  return {
    mode: value.mode === "replace" ? "replace" : "extend",
    level,
    headers: !!value.headers,
    query_params: !!value.query_params,
    req_body: !!value.req_body,
    resp_body: !!value.resp_body,
    req_body_limit: Math.max(0, Math.trunc(value.req_body_limit || 0)),
    resp_body_limit: Math.max(0, Math.trunc(value.resp_body_limit || 0))
  };
}

function cloneBasic(value?: AuthMethodBasic): AuthMethodBasic | undefined {
  if (!value) return undefined;
  return {
    users: (value.users || []).map(
      (user: AuthMethodBasicUser): AuthMethodBasicUser => ({
        username: user.username || "",
        password: user.password || ""
      })
    )
  };
}

function cloneApiKey(value?: AuthMethodApiKey): AuthMethodApiKey | undefined {
  if (!value) return undefined;
  return {
    header: value.header || "",
    keys: (value.keys || []).map(
      (item: AuthMethodApiKeyItem): AuthMethodApiKeyItem => ({
        name: item.name || "",
        key: item.key || ""
      })
    )
  };
}

function cloneJwt(value?: AuthMethodJwt): AuthMethodJwt | undefined {
  if (!value) return undefined;
  return { kid: value.kid || "", roles: [...(value.roles || [])] };
}

function cloneIpValidation(value?: AuthMethodIpValidation): AuthMethodIpValidation | undefined {
  if (!value) return undefined;
  return {
    allowed_ips: (value.allowed_ips || []).map(
      (item: AuthMethodIpValidationItem): AuthMethodIpValidationItem => ({
        name: item.name || "",
        ip: item.ip || ""
      })
    )
  };
}

export function cloneAuthMethod(value?: AuthMethod): AuthMethod {
  return {
    basic: cloneBasic(value?.basic),
    api_key: cloneApiKey(value?.api_key),
    jwt: cloneJwt(value?.jwt),
    ip_validation: cloneIpValidation(value?.ip_validation)
  };
}

export function normalizeAuth(value?: Auth | null): Auth {
  if (!value) {
    return emptyAuth();
  }
  return {
    enabled: !!value.enabled,
    mode: value.mode === "replace" ? "replace" : "extend",
    methods: (value.methods || []).map((method) => cloneAuthMethod(method))
  };
}

type RawVariables = Variable[] | Record<string, string> | null | undefined;

export function variablesToArray(variables?: RawVariables): Variable[] {
  if (!variables) {
    return [];
  }
  if (Array.isArray(variables)) {
    return variables.map((item) => ({ key: item?.key || "", value: item?.value || "" }));
  }
  return Object.entries(variables).map(([key, value]) => ({ key, value: value || "" }));
}

export function variablesToMap(variables?: RawVariables): Record<string, string> {
  const result: Record<string, string> = {};
  for (const item of variablesToArray(variables)) {
    const key = (item.key || "").trim();
    if (!key) continue;
    result[key] = item.value || "";
  }
  return result;
}

// The backend serialises variables as an object map but the form layer works
// with an ordered array; these casts bridge the two representations.
type WithRawVariables<T> = Omit<T, "variables"> & { variables?: RawVariables };

export function normalizeRoot(value: RootMain): RootMain {
  const raw = value as WithRawVariables<RootMain>;
  return {
    base_url: value?.base_url || "",
    cors: value?.cors || {
      enabled: false,
      allow_credentials: false,
      max_age: "",
      allow_origins: [],
      allow_methods: [],
      allow_headers: []
    },
    jwt: value?.jwt || [],
    auth: normalizeAuth(value?.auth),
    logging: normalizeLogging(value?.logging),
    log_own_response_errors: Boolean(value?.log_own_response_errors),
    variables: variablesToArray(raw.variables)
  };
}

export function normalizeApp(value: AppMain): AppMain {
  const raw = value as WithRawVariables<AppMain>;
  return {
    id: value?.id || "",
    active: Boolean(value?.active),
    exclude_from_metrics: Boolean(value?.exclude_from_metrics),
    path_prefix: value?.path_prefix || "",
    name: value?.name || "",
    backend: {
      url: value?.backend?.url || "",
      swagger_url: value?.backend?.swagger_url || "",
      grpc_url: value?.backend?.grpc_url || "",
      headers: value?.backend?.headers || {},
      query_params: value?.backend?.query_params || {}
    },
    auth: normalizeAuth(value?.auth),
    logging: normalizeLogging(value?.logging),
    variables: variablesToArray(raw.variables)
  };
}

export function normalizeEndpoint(value: EndpointMain): EndpointMain {
  const raw = value as WithRawVariables<EndpointMain> & {
    method?: string;
    path?: string;
    http?: { method?: string; path?: string };
  };
  return {
    id: value?.id || "",
    app_id: value?.app_id || "",
    active: Boolean(value?.active),
    exclude_from_metrics: Boolean(value?.exclude_from_metrics),
    type: value?.type === "grpc" ? "grpc" : "http",
    http: {
      method: raw.http?.method || raw.method || "",
      path: raw.http?.path || raw.path || ""
    },
    grpc: {
      service: value?.grpc?.service || "",
      method: value?.grpc?.method || "",
      path: value?.grpc?.path || ""
    },
    backend: {
      custom_path: value?.backend?.custom_path || "",
      headers: value?.backend?.headers || {},
      query_params: value?.backend?.query_params || {}
    },
    auth: normalizeAuth(value?.auth),
    logging: normalizeLogging(value?.logging),
    variables: variablesToArray(raw.variables)
  };
}

export function serializeRoot(value: RootMain): unknown {
  return { ...normalizeRoot(value), variables: variablesToMap(value?.variables) };
}

export function serializeApp(value: AppMain): unknown {
  return { ...normalizeApp(value), variables: variablesToMap(value?.variables) };
}

export function serializeEndpoint(value: EndpointMain): unknown {
  return { ...normalizeEndpoint(value), variables: variablesToMap(value?.variables) };
}
